python-for-android rebase

dbnicholson opened the pull request:
This updates the build to use the recent p4a 2022.09.04 release with our downstream changes. Mostly this is uneventful but a couple workarounds can be dropped since they've been fixed in p4a.

https://phabricator.endlessm.com/T33884

dbnicholson committed:
Update p4a

This updates the required python-for-android to the recent 2022.09.04
release. The actual version used is a `v2022.09.04-endless1` tag on our
`2022.09.04-endless` branch containing a few downstream changes and
backports.

2 changes are required in the build:

* p4a now requires NDK version 23 at a minimum.
* The bug that leaves an extra `-` in the APK and AAB filenames has been
  fixed and the renaming command needs to be updated.

https://phabricator.endlessm.com/T33884

dbnicholson committed:
Simplify zstandard patch handling

p4a 2022.09.04 includes a fix to ensure relative paths to local recipe
patches work correctly.

https://phabricator.endlessm.com/T33884

dbnicholson committed:
Drop vendored ActivityLifecycleCallbacks

This is included our 2022.09.04-endless branch and will be in the next
p4a release.

https://phabricator.endlessm.com/T33884

manuq commented on Makefile:
```diff
@@ -165,7 +165,7 @@ kolibri.apk: p4a_android_distro src/kolibri src/apps-bundle src/collections asse
 	$(P4A) apk --release --sign $(ARCH_OPTIONS) --version=$(APK_VERSION) --numeric-version=$(BUILD_NUMBER)
 	mkdir -p dist
-	mv kolibri-release-$(APK_VERSION)-.apk dist/kolibri-release-$(APK_VERSION).apk
+	mv kolibri-release-$(APK_VERSION).apk dist/kolibri-release-$(APK_VERSION).apk
```
:+1:

manuq commented on p4a-recipes/zstandard/__init__.py:
```diff
@@ -14,14 +12,7 @@ class ZStandardRecipe(CompiledComponentsPythonRecipe):
-    patches = [
-        os.path.realpath(os.path.join(os.path.dirname(__file__), "preprocessor.patch")),
-    ]
+    patches = ["preprocessor.patch"]
```
:+1:

manuq commented on src/main.py:
```diff
@@ -3,6 +3,7 @@
 import time
 
 import initialization  # noqa: F401 keep this first, to ensure we're set up for other imports
+from android.activity import register_activity_lifecycle_callbacks
```
:+1:

manuq reviewed (approved):
Looking good. Jenkins failed with:
> ERROR: Package 'python-for-android' requires a different Python: 3.6.9 not in '>=3.7.0'

dbnicholson commented:
> Looking good. Jenkins failed with:
> 
> > ERROR: Package 'python-for-android' requires a different Python: 3.6.9 not in '>=3.7.0'

Oh. I didn't notice that locally. I guess this is because the docker image is using Ubuntu bionic and that has 3.6.3. I guess I only tested in the toolbox container.

dbnicholson committed:
Update ignore rules

Several downloaded assets were missing from the gitignore rules. Update
it and sync the dockerignore rules to it to avoid copying lots of
unneeded data into the build context.

https://phabricator.endlessm.com/T33884

dbnicholson committed:
Rewrite Dockerfile to more closely resemble toolbox image

This fixes an issue where newer p4a requires python 3.7, which is not
available in Ubuntu bionic. However, it makes the build environment used
in Jenkins more closely resemble the toolbox based one that developers
use. That makes it more likely that the published build matches what
developers test locally.

Some of the system packages needed for building the app were only part
of the toolbox image as part of the eos-toolbox-dev metapackage. Those
have been added explicitly now.

https://phabricator.endlessm.com/T33884

dbnicholson commented:
> Oh. I didn't notice that locally. I guess this is because the docker image is using Ubuntu bionic and that has 3.6.3. I guess I only tested in the toolbox container.

I rewrote the non-toolbox Dockerfile so it was almost the same as the toolbox image. It worked in a local build and I think it should work in Jenkins. I like that better since the build environment in Jenkins is then closer to what we use locally.

dbnicholson committed:
Simplify kolibri wheel download

Instead of try to parse the filename from the `wget` output, just
specify a static `kolibri.whl` name to save as. The previous solution
was fragile and depended on the locale since `wget` will use fancy
quotes in UTF-8 locales.

https://phabricator.endlessm.com/T33884

manuq commented on Makefile:
```diff
@@ -61,16 +61,12 @@ clean-whl:
-	$(eval DLFILE = $(shell wget --content-disposition -P whl/ "${whl}" 2>&1 | grep "Saving to: " | sed 's/Saving to: ‘//' | sed 's/’//'))
-	$(eval WHLFILE = $(shell echo "${DLFILE}" | sed "s/\?.*//"))
-	[ "${DLFILE}" = "${WHLFILE}" ] || mv "${DLFILE}" "${WHLFILE}"
+	wget -O whl/kolibri.whl "${whl}"
```
Yes this simplify things. :+1:

manuq reviewed (commented):
I really like that the Dockerfile is closer to the toolbox now.
